processing/src/main/java/io/druid/query/filter/LikeDimFilter.java

```diff
@@ -0,0 +1,358 @@
+
+      if (suffixPattern == null && suffixStyle == SuffixStyle.MATCH_PATTERN) {
+        throw new ISE("Need suffixPattern for pattern matching!");
+      } else if (suffixPattern != null && suffixStyle != SuffixStyle.MATCH_PATTERN) {
+        throw new ISE("Can't have suffixPattern without pattern matching!");
+      }
+    }
+
+    public static LikeMatcher from(
+        final String likePattern,
+        final char escapeChar,
+        final boolean extractPrefix
+    )
+    {
+      final StringBuilder prefix = new StringBuilder();
+      final StringBuilder regex = new StringBuilder();
+      boolean escaping = false;
+      boolean inPrefix = true;
+      for (int i = 0; i < likePattern.length(); i++) {
+        final char c = likePattern.charAt(i);
+        if (c == escapeChar && escapeChar != 0) {
+          escaping = true;
+        } else if (c == '%' && !escaping) {
+          inPrefix = false;
+          regex.append(".*");
+        } else if (c == '_' && !escaping) {
+          inPrefix = false;
+          regex.append(".");
+        } else {
+          if (inPrefix && extractPrefix) {
+            prefix.append(c);
+          } else {
+            // Excessively paranoid escaping, although shouldn't affect runtime beyond compiling the regex.
+            if (DEFINITELY_FINE.matcher(String.valueOf(c)).matches()) {
+              regex.append(c);
+            } else {
+              regex.append("\\u").append(BaseEncoding.base16().encode(Chars.toByteArray(c)));
+            }
+          }
+          escaping = false;
+        }
+      }
+
+      final String regexString = regex.toString();
+
+      if (regexString.isEmpty()) {
+        return new LikeMatcher(prefix.toString(), SuffixStyle.MATCH_EMPTY, null);
+      } else if (regexString.equals(".*")) {
+        return new LikeMatcher(prefix.toString(), SuffixStyle.MATCH_ANY, null);
+      } else {
+        return new LikeMatcher(prefix.toString(), SuffixStyle.MATCH_PATTERN, Pattern.compile(regexString));
+      }
+    }
+
+    public boolean matches(@Nullable final String s)
+    {
+      if (s == null) {
+        return prefix == null && (suffixStyle == SuffixStyle.MATCH_ANY
+                                  || suffixStyle == SuffixStyle.MATCH_EMPTY
+                                  || suffixPattern.matcher("").matches());
+      }
+
+      final String suffix;
+
+      if (prefix == null) {
+        suffix = s;
+      } else if (s.startsWith(prefix)) {
+        suffix = s.substring(prefix.length());
+      } else {
+        return false;
+      }
+
+      if (suffixStyle == SuffixStyle.MATCH_ANY) {
+        return true;
+      } else if (suffixStyle == SuffixStyle.MATCH_EMPTY) {
+        return suffix.isEmpty();
+      } else {
+        // suffixStyle is MATCH_PATTERN
+        return suffixPattern.matcher(suffix).matches();
+      }
+    }
+
+    public DruidPredicateFactory predicateFactory(final ExtractionFn extractionFn)
+    {
+      return new DruidPredicateFactory()
```

Comment:
IMO the anonymous class is easier to read (much less boilerplate code). My personal feeling is the balance between code readability and logging readability is in favor of anonymous classes for things like this. But I'm open to being convinced otherwise or outvoted.